### ISSUE TITLE
Git Clone does not work with Azure DevOps repos

### ISSUE THREAD

**Issue by cplaetzinger** [go-git/go-git#64]
I'm trying to do a git clone on a git repo hosted on Azure DevOps. This fails with "error: unexpected client error: unexpected requesting "https://PATH-TO-REPO/git-upload-pack" status code: 400"

Seems to be the same error as in https://github.com/src-d/go-git/issues/1058

If you need more information please let me know

**Comment by mcuadros** [comment]
Yes, the problem still exists.

**Comment by artbegolli** [comment]
cry

**Comment by artbegolli** [comment]
Any update on the git v2 protocol implementation?

**Comment by mgibson1121** [comment]
Hello, 

The [kpack](https://github.com/pivotal/kpack) team is seeing this problem with our users as well when they try to integrate with Azure DevOps.  Just surfacing in case it helps with feature prioritization.

The error in the build logs is this:

Loading secrets for "https://dev.azure.com" from secret "azure-devops-git"
unable to fetch git repository: unexpected client error: unexpected requesting "https://dev.azure.com/*******/" status code: 400

**Comment by phillebaba** [comment]
So I have been doing some research into this during the last two weekends, and thought I could share what I have found.

TLDR; The reason we are seeing error messages has less to do with git v2 protocol and more to do with Azure DevOps requiring clients to support multi-ack. Nothing new really going by [this issue](https://github.com/src-d/go-git/issues/335#issuecomment-310791062) in the "old" repo.

I followed the suggestions in the issue and attempted to just get "dumb" multi-ack working, which I seem to have achieved with worryingly small amounts of code changes. Most components were there, I just needed to take the changes made in @bashims [PR](https://github.com/go-git/go-git/pull/111) to get external refs working. Here is my fork that currently works with Azure DevOps.
https://github.com/phillebaba/go-git

I have created an example project if you want to try this out yourself. The notification-controller repo has been imported from fluxcd/notification-controller to have way to test identical repos in GitHub and Azure DevOps. You obviously need to set up your own repos as git over ssh requires authenticated users.
https://github.com/phillebaba/go-git-demo
```
go run main.go --url https://dev.azure.com/philiplaine0831/TEST/_git/notification-controller --destination /tmp/http-azdo-noti
ls /tmp/http-azdo-noti
go run main.go --url git@ssh.dev.azure.com:v3/philiplaine0831/TEST/notification-controller --ssh-key-path ~/.ssh/id_rsa --destination /tmp/azdo-noti
ls /tmp/azdo-noti
go run main.go --url git@github.com:phillebaba/notification-controller.git --ssh-key-path ~/.ssh/id_rsa --destination /tmp/gh-noti
ls /tmp/gh-noti
```

There is one detail left however, SSH technically works but not without a small cheat.
https://github.com/phillebaba/go-git/blob/3fa0eaceaaf7d66ff559278afc3090e8d6eb14cd/plumbing/transport/internal/common/common.go#L361
Without commenting out this line of code you get an error when closing the connection in the end. The error only appears when cloning larger repositories, and only on Azure DevOps. My guess is that the server is closing the connection before the client can, but I am not too sure. It doesn't help that debugging ssh connections is a pain.
 ```
close tcp 10.134.10.7:45060->40.74.28.11:22: use of closed network connection
```

I would appreciate if anyone could verify the work I have done, as I am pretty new to how the git protocols work. My hope is that we could make an initial PR with the "dumb" multi-ack to get go-git working with Azure DevOps. Then we could attempt implementing proper multi-ack logic which is a bit more complicated.

**Comment by bashims** [comment]
@phillebaba it looks like my PR is blocked on the test coverage check. I will need to get some pack files to test against.

**Comment by phillebaba** [comment]
@bashims are the tests something that you are actively working on, or do you need help getting that done?

**Comment by bashims** [comment]
@phillebaba, yes I should have them wrapped up by tomorrow.

**Comment by roberth1988** [comment]
Hi guys ;) any update on this issue? I've stumbled upon that problem too :/ - blocks me from using go-git on Azure DevOps.

**Comment by ondrejkolin** [comment]
Seems to be a blocker to different projects. I've experienced this issue with flux-cli tool. Any updates on this topic?

**Comment by phillebaba** [comment]
@ondrejkolin this has been for the most part solved in Flux as we have moved to support libgit2 as an alternative. I am doubtful we will get support in go-git unless someone decides its worth to spend the time. I had a stab at it 6 months ago and came up with a partial solution which would not be accepted in go-git. It would take someone to fully implement multi-ack capabilities into go-git which isn't the easiest thing to do.

**Comment by sluongng** [comment]
Just FYI https://github.com/google/gitprotocolio has done most of the hard work of parsing git protocol v2.

https://github.com/google/goblet provide a sample usage of that library.

@mcuadros I think that should simplify matters by quite a bit.

**Comment by jdswed** [comment]
Is there an update on getting multi-ack?

**Comment by bluebrown** [comment]
what a bummer, There is quite a few of these with this project it seems.

**Comment by pjbgf** [comment]
Once merged, this PR will enable clones from Azure DevOps: https://github.com/go-git/go-git/pull/613.

**Comment by pjbgf** [comment]
The PR above has been merged, so give it a try based on [_examples/azure_devops](https://github.com/go-git/go-git/blob/master/_examples/azure_devops/main.go).

**Comment by bluebrown** [comment]
@pjbgf , I see in the example that it says fetches are problematic. Does this include pull? Because I would use this for a git bot which needs to refresh the repo frequently.

**Comment by cherrymu** [comment]
Hey @pjbgf, is the fix for this issue inclued only in **go-git/v5** version only and not before?

**Comment by pjbgf** [comment]
@cherrymu the fix is part of master, and not on a tagged release. To update to the latest in master you can:
`go get github.com/go-git/go-git/v5@master`

**Comment by bluebrown** [comment]
@pjbgf , I tried to follow your optimization technique, but the remote HEADs  hash comes as `0000000000000000000000000000000000000000`. Therefore, it can't be compared to local. Any idea why that is?


<details>
<summary>Code Snippet</summary>

```go
w, err := t.repo.Worktree()
if err != nil {
	return fmt.Errorf("could not refresh: could not get worktree: %w", err)
}

if err := w.Checkout(&git.CheckoutOptions{
	Branch: plumbing.ReferenceName(fmt.Sprintf("refs/heads/%s", branch)),
	Force:  true,
}); err != nil {
	return fmt.Errorf("could not refresh: could not checkout: %w", err)
}

h, err := t.repo.Head()
if err != nil {
	return fmt.Errorf("could not refesh: could not get head: %w", err)
}

remote, err := t.repo.Remote("origin")
if err != nil {
	return fmt.Errorf("could not refesh: could not get remote: %w", err)
}

refs, err := remote.ListContext(ctx, &git.ListOptions{
	Auth: t.gauth,
})
if err != nil {
	return fmt.Errorf("could not refesh: could not get remote-list: %w", err)
}

// check if the current heads hash is the same as the remote hash
// if they are the same, return ealry to avoid doing and unneccary clone
// this is not required anymore if multi ack is completely supported by
// go-git, since then a fetch/pull is possible
for _, r := range refs {
	if r.Name() == "HEAD" {
		log.Trace().Str("remote", r.Hash().String()).Str("local", h.Hash().String()).Msg("head refs")
		if r.Hash() == h.Hash() {
			// nothing to do
			return nil
		}
	}
}
```

</details>

**Comment by anubhav06** [comment]
Hey @pjbgf . While cloning a **public** Azure repo, I get an error `authentication required`.
If I pass my Personal Access Token (PAT), then it clones the repo, however without the PAT, it is not able to clone, even though the repo is public

**Comment by pjbgf** [comment]
@anubhav06 can you please create a new issue just focusing on public Azure DevOps repositories? It would be great if you could share the code snippet with the repository you are trying to access.

**Comment by pjbgf** [comment]
@bluebrown one of the limitations of the approach is that the repository can only do a single handshake with the remote - on do follow-up actions via `ListContext`. The optimisation I suggested, assumes the list operation taking place on a completely new repository. You then use the result to compare the commit ID with the commit ID from your existing repository.

**Comment by JanhaviAlekar** [comment]
Does this issue still exist? Or does go-git support Azure Repositories now?
If not ,can anyone suggest alternative for it

**Comment by amirschw** [comment]
It is supported for Azure Repos today

**Comment by baterja** [comment]
>It is supported for Azure Repos today

Huh, since which version? Why isn't it closed then?

**Comment by pjbgf** [comment]
Clone of Azure DevOps repositories is supported with the caveats defined in https://github.com/go-git/go-git/pull/613#issue-1451371531 since version [v5.5.0](https://github.com/go-git/go-git/releases/tag/v5.5.0). A working example is available within [_examples/azure_devops](https://github.com/go-git/go-git/blob/master/_examples/azure_devops/main.go).

Closing this to avoid future confusion. For additional support please open a new issue.

**Comment by pjbgf** [comment]
@anubhav06 for future reference, public repositories work fine as well:

`go run _examples/azure_devops/main.go https://dev.azure.com/public-git/sample/_git/sample /tmp/sample "" ""`

**Comment by kjwhal-osh** [comment]
Hey all,

RedPanda Console is on 5.13 (https://github.com/redpanda-data/console/blob/d6119efd7582ad457ab1fc2ccbec84cf4385dd42/backend/go.mod#L23)

And I'm seeing the same error as described in the initial post

.../git-upload-pack\" status code: 400"}   

I've verified the PAT used for access is good, and I'm not seeing problems in our istio egress

Is there another way to troubleshoot this?
